base/strings/string.jl

```diff
@@ -190,14 +191,135 @@ end
+                    (3,  12), (4,  1), (3,  2), (11, 1), (6, 3), (5,  1), (8, 11)]
+
+    # These are the rows discussed in comments above
+    state_arrays = [[ 0, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 1, 1, 0, 2, 1, 2, 1, 3, 3, 1],
+                    [ 2, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 3, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 5, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 8, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 7, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 1, 1, 0, 2, 2, 1, 3, 3, 1, 1],
+                    [ 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 1, 1, 0, 2, 1, 2, 3, 3, 1, 1],
+                    [ 4, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+                    [ 6, 1, 1, 1, 1, 1, 1, 1, 1, 1]]
+
+    #This converts the state_arrays into the shift encoded UInt64
+    class_row = zeros(UInt64, num_classes)
+    for i = 1:num_classes
+        row = UInt64(0)
+        for j in 1:num_states
+            #Calculate the shift required for the next state
+            to_shift = UInt8((state_arrays[i][j]) * bit_per_state)
+            #Shift the next state into the position of the current state
+            row = row | (UInt64(to_shift) << ((j - 1) * bit_per_state))
+        end
+        class_row[i]=row
+    end
+    mapreduce(t->fill(class_row[t[1]+1],t[2]),vcat,class_repeats)
+end
+
 
-byte_string_classify(s::Union{String,Vector{UInt8},FastContiguousSubArray{UInt8,1,Vector{UInt8}}}) =
-    ccall(:u8_isvalid, Int32, (Ptr{UInt8}, Int), s, sizeof(s))
+const _UTF8_DFA_ACCEPT = UInt64(0) #This state represents the start and end of any valid string
+const _UTF8_DFA_INVALID = UInt64(6) # If the state machine is ever in this state just stop
+
+# This function is designed so that you could use it on strings with discontinous memmory layouts
+# by only feeding it contiguous block and keeping track of the state inbetween.
+# Furthermore you could check in returned value is _UTF8_DFA_INVALID and stop as invalid if it was.
+function _isvalid_utf8_dfa(bytes::Vector{UInt8},state::UInt64 = _UTF8_DFA_ACCEPT)
+    f(byte) = @inbounds _UTF8_DFA_TABLE[byte+1]
+    op(s, byte_dfa) = byte_dfa >> (s & UInt64(63))
+    final_state = mapfoldl(f, op, bytes, init = state)
+    return (final_state & UInt64(63))
+end
+
+# This is a shift based utf-8 DFA that works on string that are a contiguous block
+function _isvalid_utf8(bytes::Vector{UInt8})
+    final_state = _isvalid_utf8_dfa(bytes, _UTF8_DFA_ACCEPT)
+    return (final_state & UInt64(63)) == _UTF8_DFA_ACCEPT
+end
+
+_isvalid_utf8(s::Union{String,FastContiguousSubArray{UInt8,1,Vector{UInt8}}}) = _isvalid_utf8(unsafe_wrap(Vector{UInt8}, s))
+
+# Classifcations of string
     # 0: neither valid ASCII nor UTF-8
     # 1: valid ASCII
     # 2: valid UTF-8
+function byte_string_classify(s::Union{String,FastContiguousSubArray{UInt8,1,Vector{UInt8}}})
+    bytes = unsafe_wrap(Vector{UInt8}, s)
+    byte_string_classify(bytes, kwargs...)
```

Comment:
Does this need a `GC.@preserve s`?  (Similarly below.) i.e.
```jl
GC.@preserve s byte_string_classify(unsafe_wrap(Vector{UInt8}, s), kwargs...)
```

Reply:
Changed to using `CodeUnits`